docs(cardinal): adds documentation for CQL to world-engine. (#266)
Argus-Labs/world-engine: docs/pages/Cardinal/API-Reference/cql.mdx

> # Component Query Language

Reviewer: In the preview link, this page appears as "Cql" in the sidebar, how can we change it to be "CQL" instead?

Author: You mean the file name?

Reviewer 2: the actual item in the side bar here: https://world-engine-docs-2njn8wjng-argus-labs.vercel.app/

it's showing as `Cql`. youll have to edit the _meta.json

## Language Description:

CQL (Component Query Language) is a simple query language that queries entities based on components contained in the entity.

> In the language, you have components, which is simply the names of the specified components.

Reviewer 2: ```suggestion
The language uses the names of `ComponentType`s to specify the entities you are interested in querying.
```

Examples:

* healthComponent
* attackComponent
* armComponent
* legComponent

> You also have two functions which can be called with a variadic amount of components as parameters:

Reviewer 2: ```suggestion
Two functions are provided in the language that accept a variadic amount of components as parameters
```

> `EXACT(…) and CONTAINS(…)`

Reviewer 2: maybe here we can add a header, Functions, and each sub-header can be the function types, contains/exact, followed by description + examples

CONTAINS is a query for entities that only need to "contain" the specified components. If they have more components then just the specified components then it is still a valid entity for that query.

EXACT is a query for entities that contain "exactly" the specified the components. Nothing more nothing less.

Examples:

- `EXACT(healthComponent)` is a query for all entities with exactly one health component nothing more.
- `EXACT(healthComponent, attackComponent)` is a query for all entities with exactly one health component and one attack component, nothing more.
- `CONTAINS(armComponent)` is a query for all entities that have a arm component. The entity can have more components then just the arm.
- `CONTAINS(armComponent, legComponent)` is query for all entities that have both an arm component and a leg component. The entity can have more components then the arm and the leg.

> In CQL there three operators: `&`, `|` and `!`

Reviewer 2: maybe make a header for logical operators here, same as above

`|` performs the "or" operation on two queries.
- Example: `CONTAINS(healthComponent) | !CONTAINS(attackComponent)` is a query for all entities that either contains a health component "or" does not contain an attack component

Operators in CQL do not have any intrinsic precedence. All expressions are consumed from left to right.
- Example: `A & B | C & D | F`  is equivalent to `( ( A & B ) | C ) & D ) | F)`

You can use parenthesis to specify and change precedence in CQL.
- Example: `EXACT(legComponent) | (!CONTAIN(healthComponent) & !CONTAIN(attackComponent))`
  - The above is a query for either an entity with only a leg component or an entity that does not have a health component and also does not have an attack component.

- Example: `(EXACT(legComponent) | !CONTAIN(healthComponent)) & !CONTAIN(attackComponent)`
  - The above is the same query but with precedence changed. Now it is querying an entity with either exactly one leg component or does not have a have a health component. Additionally that entity must not ever contain a attack component.

## How to use CQL:

> There is a http route on Cardinal to use CQL. To execute a query use a POST request to the url with path `query/game/cql`. In the body place a json object in the format: `{"CQL": "<CQL as a string>"}` The results of the query will be an array of objects with an entity id and another nested array of component data relevant to the entity id.

Reviewer 2: i think this part could be a bit more aligned with other API documentation where you have a spot for path, request type, response type, and examples, rather than having this all in one paragraph

# Component Query Language

## Language Description

CQL (Component Query Language) is a simple query language that queries entities based on components contained in the entity.

### Components

> The language uses the names of `ComponentType`s to specify the entities you are interested in querying.

Reviewer 2: on second thought, its probably better if we just link to components here:

```suggestion
import Link from 'next/link'
The language uses the names of <Link href="/Cardinal/API-Reference/Components">Components</Link> to specify the entities you are interested in querying.
```
